atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java

```diff
@@ -214,7 +216,47 @@ public CompletableFuture<JoinResponse> onJoin(final JoinRequest request) {
                     .withError(Type.PROTOCOL_ERROR, throwable.getMessage())
                     .build();
               }
-              return JoinResponse.builder().withStatus(Status.OK).build();
+              if (reconfigureResponse.status() == Status.OK) {
+                return JoinResponse.builder().withStatus(Status.OK).build();
+              } else {
+                return JoinResponse.builder()
+                    .withStatus(Status.ERROR)
+                    .withError(reconfigureResponse.error())
+                    .build();
+              }
+            });
+  }
+
+  @Override
+  public CompletableFuture<LeaveResponse> onLeave(final LeaveRequest request) {
+    raft.checkThread();
+    final var currentConfiguration = raft.getCluster().getConfiguration();
+    final var updatedMembers =
```

Comment:
:wrench: Here we can return immediately if the member is not in the configuration, meaning it has already left. There is no need to go through the joint-consensus again.

Reply:
I had something like this for a while but then I tripped over that `DefaulRaftMember#equals` is only comparing ids, not type and I accidentally broke changing the type of a member. I implemented it again as part of `onReconfigure` where we now just do nothing if member ids and types didn't change.